src/mono/mono/mini/interp/interpgo.c

```diff
@@ -0,0 +1,395 @@
+        case 9:  k2 ^= (uint64_t)(tail[8]) << 0;
+            k2 *= c2; k2 = ROTL64(k2, 33); k2 *= c1; h2 ^= k2;
+
+        case 8: k1 ^= (uint64_t)(tail[7]) << 56;
+        case 7: k1 ^= (uint64_t)(tail[6]) << 48;
+        case 6: k1 ^= (uint64_t)(tail[5]) << 40;
+        case 5: k1 ^= (uint64_t)(tail[4]) << 32;
+        case 4: k1 ^= (uint64_t)(tail[3]) << 24;
+        case 3: k1 ^= (uint64_t)(tail[2]) << 16;
+        case 2: k1 ^= (uint64_t)(tail[1]) << 8;
+        case 1: k1 ^= (uint64_t)(tail[0]) << 0;
+            k1 *= c1; k1 = ROTL64(k1, 31); k1 *= c2; h1 ^= k1;
+            break;
+    }
+
+    // finalization
+    h1 ^= len;
+    h2 ^= len;
+    h1 += h2;
+    h2 += h1;
+    h1 = fmix64(h1);
+    h2 = fmix64(h2);
+    h1 += h2;
+    h2 += h1;
+
+    setblock64((uint8_t*)(out), h1);
+    setblock64((uint8_t*)(out) + sizeof(uint64_t), h2);
+}
+
+// end of murmurhash
+
+
+typedef struct {
+	uint8_t *data;
+	uint32_t size, capacity;
+	mono_mutex_t mutex;
+} interpgo_table;
+
+// loaded_table is the table we loaded from persistent storage at startup (if any),
+//  while building_table is the table we built during the current run. we store these
+//  separately so that we don't have to maintain a sorted table (for bsearch) on an
+//  ongoing basis.
+static interpgo_table *loaded_table = NULL, *building_table = NULL;
+
+static int
+hash_comparer (const void *needle, const void *haystack)
+{
+	return memcmp (needle, haystack, MM3_HASH_BYTE_SIZE);
+}
+
+static gboolean
+table_lookup (interpgo_table *table, uint8_t hash[MM3_HASH_BYTE_SIZE]) {
+	// Early out if no table is loaded or the table is empty.
+	if (!table || !table->size)
+		return FALSE;
+
+	g_assert (table->size <= table->capacity);
+
+	// FIXME: We need to perform this search inside the lock because another thread could
+	//  load a new interpgo table during the search and then attempt to g_free the old table's
+	//  data out from under us, and we would crash. This means that multiple threads can't
+	//  perform a search in this table at once. It would be nice if we had a rwlock for this
+	// In disable_threads builds and situations where no table is loaded this doesn't matter
+	// Another solution would be to make the buffer ref-counted, so that it only gets freed
+	//  once nobody is using it. But that's worse in the no concurrent access case since
+	//  we'd waste cycles modifying the refcount when it doesn't matter
+	mono_os_mutex_lock (&table->mutex);
+	void * result = mono_binary_search (hash, table->data, table->size / MM3_HASH_BYTE_SIZE, MM3_HASH_BYTE_SIZE, hash_comparer);
+	mono_os_mutex_unlock (&table->mutex);
+	return (result != NULL);
+}
+
+static void
+table_add (interpgo_table **table_variable, uint8_t hash[MM3_HASH_BYTE_SIZE]) {
+	interpgo_table *table = *table_variable, *temp;
+	// If we don't have a table yet, allocate one
+	if (!table) {
+		table = g_malloc0 (sizeof (interpgo_table));
+		// We may have lost a race against another thread that just allocated a table
```

Comment:
table->mutex is not initialized ?

Reply:
i've never seen any code initialize our mutexes. size/capacity are initialized on-demand later guarded by the mutex